.github/workflows/build-and-test-toolchain.yml

```diff
@@ -0,0 +1,139 @@
+on:
+  workflow_dispatch:
+    inputs:
+      binutils_branch:
+        description: 'Binutils branch to build'
+        required: false
+        default: 'woarm64'
+      gcc_branch:
+        description: 'GCC branch to build'
+        required: false
+        default: 'woarm64'
+      mingw_branch:
+        description: 'MinGW branch to build'
+        required: false
+        default: 'woarm64'
+      arch:
+        description: 'Architecture to build for'
+        required: false
+        default: 'x86_64'
+      platform:
+        description: 'Platform to build for'
+        required: false
+        default: 'w64-mingw32'
+      crt:
+        description: 'C runtime to build for'
+        required: false
+        default: 'msvcrt'
+      tag:
+        description: 'Tag to use for the artifact'
+        required: true
+  workflow_call:
+    inputs:
+      binutils_branch:
+        type: string
+      gcc_branch:
+        type: string
+      mingw_branch:
+        type: string
+      arch:
+        type: string
+      platform:
+        type: string
+      crt:
+        type: string
+      tag:
+        type: string
+
+env:
+  BINUTILS_REPO: Windows-on-ARM-Experiments/binutils-woarm64
+  BINUTILS_BRANCH: ${{ inputs.binutils_branch || 'woarm64' }}
+  BINUTILS_VERSION: binutils
+
+  GCC_REPO: Windows-on-ARM-Experiments/gcc-woarm64
+  GCC_BRANCH: ${{ inputs.gcc_branch || 'woarm64' }}
+  GCC_VERSION: gcc
+
+  MINGW_REPO: Windows-on-ARM-Experiments/mingw-woarm64
+  MINGW_BRANCH: ${{ inputs.mingw_branch || 'woarm64' }}
+  MINGW_VERSION: mingw-w64
+
+  ARCH: ${{ inputs.arch || 'x86_64' }}
+  PLATFORM: ${{ inputs.platform || 'w64-mingw32' }}
+  CRT: ${{ inputs.crt || 'msvcrt' }}
+
+  SOURCE_PATH: ~/code
+  DOWNLOADS_PATH: ~/downloads
+  ARTIFACT_PATH: ~/artifacts
+
+jobs:
+  build-and-test-toolchain:
+    name: Build and test toolchain
+    runs-on: windows-latest
+
+    defaults:
+      run:
+        shell: wsl-bash {0}
+
+    steps:
+      - name: Setup WSL
+        uses: Vampire/setup-wsl@v3
+        with:
+          distribution: Ubuntu-22.04
+          additional-packages:
+            git
+
+      - name: Checkout repository
+        run: |
+          git clone --depth=1 https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build.git -b ${{ github.head_ref }} .
+
+      - name: Checkout binutils
+        run: |
+          git clone --depth=1 https://github.com/${{ env.BINUTILS_REPO }}.git -b ${{ env.BINUTILS_BRANCH }} ${{ env.SOURCE_PATH }}/${{ env.BINUTILS_VERSION }}
+
+      - name: Checkout GCC
+        run: |
+          git clone --depth=1 https://github.com/${{ env.GCC_REPO }}.git -b ${{ env.GCC_BRANCH }} ${{ env.SOURCE_PATH }}/${{ env.GCC_VERSION }}
+
+      - name: Checkout MinGW
+        run: |
+          git clone --depth=1 https://github.com/${{ env.MINGW_REPO }}.git -b ${{ env.MINGW_BRANCH }} ${{ env.SOURCE_PATH }}/${{ env.MINGW_VERSION }}
+
+      - name: Build toolchain
+        run: |
+          export ARCH=${{ env.ARCH }}
```

Comment:
I am not happy about this but I haven't find a better solution how to transfer environment variables from host to WSL.